src/Ouroboros/Network/Protocol/UntypedStream/Client.hs

```diff
@@ -0,0 +1,107 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module Ouroboros.Network.Protocol.UntypedStream.Client where
+
+import Control.Monad (void)
+import Data.Word (Word32)
+import Protocol.Untyped
+
+import Ouroboros.Network.Protocol.UntypedStream.Type
+import Ouroboros.Network.MonadClass
+
+data UntypedStreamClient m range a t where
+  SendMsgRequest
+    :: range
+    -> Window
+    -> Threshold
+    -> ClientHandleData m a t
+    -> UntypedStreamClient m range a t
+
+data ClientHandleData m a t =
+  ClientHandleData {
+    recvMsgData        :: a -> m (ClientHandleData m a t),
+    recvMsgStreamEnd   :: m t
+  }
+
+-- | Consumer part of the block streaming protocol.
+--
+-- When the consumer will receive `Threshold` messages it will send
+-- `MsgUpdateWindow` with the difference between window and the threshold.
+--
+--  In the following digrams, lines from client to server represent
+--  `MsgRequestBlock` (the initial line), or `MsgUpdateWindow`.  The lines from
+--  server to client represent `MsgStreamBlock`.
+--
+--  `x` represent the point when the client will block (not sending next
+--  `MsgUpdateQueue`).  This can happen after receiving `window + (window
+--  - threshold)` messages.  This is the maximum number of allocations the
+--  client must be prepared for.
+--
+--  window    = 3
+--  threshold = 2
+--
+--  server -----------------------
+--            / \ \ \  /\    /\
+--           /   \ \ \/  \  /  \
+--          /     \ \/\   \/    \
+--  client ---------------x-------
+--
+--  window    = 4
+--  threshold = 2
+--
+--  server -----------------------
+--            / \ \ \ \/\ \  /\ \
+--           /   \ \ \/\ \ \/  \ \
+--          /     \ \/\ \ \/\   \ \
+--  client ---------------x-------
```

Comment:
What I had in mind for the block download protocol was to include in each request for blocks a number meaning "give me this many". The client presumably knows how many blocks are outstanding, since the chain sync protocol gives block counts for the tip and read pointer. It also knows the hashes of the read pointer and tip. Using the hashes, and block-count offset and range, a client can request any contiguous section of the chain between the read pointer and tip. Instead of updating a window, it can simply send a new request while one is outstanding, with the offset at the end of the prior request.

Reply:
Using this synchronous protocol, which can be expressed in the typed framework:

```
data TrBlockDownload from to where
  TrRequest :: HeaderHash -> HeaderHash -> Offset -> Range -> TrBlockDownload 'StIdle 'StBusy
  TrDeliver :: HeaderHash -> Offset -> Body -> TrBlockDownload 'StBusy 'StBusy
  TrEnd :: TrBlockDownload 'StBusy 'StTerminal
```

To get latency hiding, we can run instance of this protocol concurrently, multiplexed over the same channel. When one instance has received, say, half of the range requested, a new instance can be spun up requesting another range, ending at the end of the other instance's range.

Reply:
Sure, that's an option which I considered.  I went this path only because I hoped for latency hiding version of typed transitions.